### PR TITLE
Placeholder dashboards

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Add placeholder dashboard(s) which show some fake data for a more attractive default dashboard.

### CODE REVIEW

File: src/Dashboard/Grid.php [hunk 1]
```diff
@@ -380,6 +382,23 @@ public function show(bool $mini = false, ?string $token = null)
...
+            ];
+            $filters = TemplateRenderer::getInstance()->renderFromStringTemplate(<<<TWIG
+                <div class='placeholder_info' style="background-color: transparent; color: var(--tblr-body-color); font-size: var(--tblr-body-font-size)">
+                    <div class="alert alert-warning">{{ placeholder_msg }}</div>
+                    <button class="btn btn-primary btn-sm change-global-dashboard"
+                            onclick="window.location='{{ change_global_url }}'" role="link">{{ change_global_msg }}</button>
+                    <button class="btn btn-primary btn-sm change-my-dashboard"
+                            onclick="window.location='{{ change_my_url }}'" role="link">{{ change_my_msg }}</button>
+                </div>
+TWIG, $params);
```
Comment (orthagh) [inline src/Dashboard/Grid.php]:
```suggestion
            $params = [
                'change_global_url' => \Config::getFormURL() . '?forcetab=Config$2',
                'change_my_url' => \Preference::getSearchURL() . '?forcetab=Config$1'
            ];
            $filters = TemplateRenderer::getInstance()->renderFromStringTemplate(<<<TWIG
                <div class='placeholder_info' style="background-color: transparent; color: var(--tblr-body-color); font-size: var(--tblr-body-font-size)">
                    <div class="alert alert-info m-2">
                        <div class="d-flex">
                            <i class="ti ti-info-circle fa-2x me-3"></i>
                            <div>
                                <h4 class="alert-title">
                                    {{ __('This is a placeholder dashboard.') }}
                                </h4>
                                <div class="text-secondary">
                                    {{ __(' You can switch to a dashboard with your real data in the general setup or your user preferences.') }}
                                </div>

                                <div class="mt-2">
                                    <a class="btn btn-info change-global-dashboard me-2"
                                    href="{{ change_global_url }}">
                                        <i class="ti ti-settings-down"></i>
                                        <span>{{ __('Change default dashboard') }}</span>
                                    </a>
                                    <a class="btn btn-info change-my-dashboard"
                                    href="{{ change_my_url }}">
                                        <i class="ti ti-user-cog"></i>
                                        <span>{{ __('Change my default dashboard') }}</span>
                                    </a>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
TWIG, $params);
```

changes:
- use a more complex alert layout (title, sub text, "footer")
- use info actions class (instead the mix of warning/primary)
- add icons
- some spacing
- use href instead onclick+window.location (didn't understand this one)
- use inline gettext (not using passing params) in inline twig

image

**Comment by orthagh** [comment]
One thing for the links, an anchor pointing on the good setting will be welcomed.
Even better, an highlight of the label option after anchor scrolling.

File: src/Dashboard/Grid.php [hunk 2]
```diff
@@ -370,16 +371,43 @@ public function show(bool $mini = false, ?string $token = null)
...
-         </div>
-HTML;
+            $params = [
+                'change_global_url' => \Config::getFormURL() . '?forcetab=Config$2#section-dashboards',
+                'change_my_url' => \Preference::getSearchURL() . '?forcetab=Config$1#section-dashboards',
+                'filter_toolbar_class' => $placeholder ? 'd-none' : '',
+                'is_placeholder' => $placeholder ? 1 : 0
+            ];
+            $filters = TemplateRenderer::getInstance()->renderFromStringTemplate(<<<TWIG
+                <div class='filters_toolbar {{ filter_toolbar_class }}'>
```
Comment (orthagh) [inline src/Dashboard/Grid.php:381]:
```suggestion
                <div class='filters_toolbar m-2 {{ filter_toolbar_class }}'>
```

File: src/Dashboard/Grid.php [hunk 3]
```diff
@@ -370,16 +371,43 @@ public function show(bool $mini = false, ?string $token = null)
...
+                    </span>
+                </div>
+                <div class='placeholder_info {{ is_placeholder ? "" : "d-none" }}' style="background-color: transparent; color: var(--tblr-body-color); font-size: var(--tblr-body-font-size)">
+                    <div class="alert alert-info">
+                        <div class="d-flex">
+                            <i class="ti ti-info-circle fa-2x me-3"></i>
+                            <div>
+                                <h4 class="alert-title">{{ __('This is a placeholder dashboard.') }}</h4>
+                                <div class="text-secondary">{{ __(' You can switch to a dashboard with your real data in the general setup or your user preferences.') }}</div>
+                                <div class="mt-2>">
```
Comment (orthagh) [inline src/Dashboard/Grid.php:396]:
```suggestion
                                <div class="mt-2">
```

File: templates/pages/setup/general/base_form.html.twig [hunk 4]
```diff
@@ -54,3 +54,27 @@
         </div>
     </form>
 {% endif %}
+
+<script>
+    $(() => {
+        // Handle anchor in URL if present since this content is loaded via AJAX
+        if (window.location.hash && $(window.location.hash).length > 0) {
```
Comment (orthagh) [inline templates/pages/setup/general/base_form.html.twig:61]: Is the code in `src/Ajax.php` (l386) doesn't work already ? It seems like a doublon for me

**Review by orthagh — CHANGES_REQUESTED** [review]

File: templates/pages/setup/general/base_form.html.twig (hunk 4 above)
Comment (cconard96) [reply templates/pages/setup/general/base_form.html.twig:61]: No. It seems like the auto-scroll and animation are only triggered when you switch tabs and doesn't work for the initial tab load. If that is resolved, then I can probably remove the code for it in this PR.

Comment (orthagh) [reply templates/pages/setup/general/base_form.html.twig]: pushed a commit for that, could you check if it's ok for you

File: install/empty_data.php [hunk 5]
```diff
@@ -940,6 +941,37 @@ public function getEmptyData(): array
             $i++;
         }
 
+        // Placeholder dashboards
+        $tables['glpi_dashboards_dashboards'][] = [
+            'id' => $i,
+            'is_placeholder' => 1,
+            'key' => 'central_placeholder',
+            'name' => 'Example central',
```
Comment (cedric-anne) [inline install/empty_data.php:949]: I think it should be translated.
```suggestion
            'name' => __('Example central'),
```

File: src/Dashboard/Grid.php [hunk 6]
```diff
@@ -370,16 +371,43 @@ public function show(bool $mini = false, ?string $token = null)
...
-         </div>
-HTML;
+            $params = [
+                'change_global_url' => \Config::getFormURL() . '?forcetab=Config$2#section-dashboards',
+                'change_my_url' => \Preference::getSearchURL() . '?forcetab=Config$1#section-dashboards',
+                'filter_toolbar_class' => $placeholder ? 'd-none' : '',
+                'is_placeholder' => $placeholder ? 1 : 0
+            ];
+            $filters = TemplateRenderer::getInstance()->renderFromStringTemplate(<<<TWIG
+                <div class='filters_toolbar m-2 {{ filter_toolbar_class }}'>
```
Comment (cedric-anne) [inline src/Dashboard/Grid.php:381]: No need to have a specific variable passed to template for this.
```suggestion
                <div class='filters_toolbar m-2 {{ is_placeholder ? "d-none" : "" }}'>
```

File: src/Dashboard/Grid.php [hunk 7]
```diff
@@ -370,16 +371,43 @@ public function show(bool $mini = false, ?string $token = null)
...
-            </span>
-         </div>
-HTML;
+            $params = [
+                'change_global_url' => \Config::getFormURL() . '?forcetab=Config$2#section-dashboards',
+                'change_my_url' => \Preference::getSearchURL() . '?forcetab=Config$1#section-dashboards',
+                'filter_toolbar_class' => $placeholder ? 'd-none' : '',
+                'is_placeholder' => $placeholder ? 1 : 0
+            ];
+            $filters = TemplateRenderer::getInstance()->renderFromStringTemplate(<<<TWIG
```
Comment (cedric-anne) [inline src/Dashboard/Grid.php:380]: The problem with usage of `renderFromStringTemplate()` is that usages of translation function (e.g. `__()`) will not be detected by `xgettext` as they are located in a heredoc.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Comment by orthagh** [comment]
We may move from the image to a class provider giving "real" fake date and make a real dashboard instead.
Like discussed together, image doens't scale well with the several screen size glpi can be used.

**Comment by orthagh** [comment]
> Hard-coded values

This is the way. (can also be tested later with panther)

**Comment by cconard96** [comment]
Some changes:
1. A placeholder version of every default dashboard is now created during install and made the default.
2. Placeholder dashboards can have any card added to them and they will reflect fake data. None of the data is random, but some of it is calculated based on static values in order to emulate a larger dataset.
3. Single line graphs seemed broken (not related to this PR). I added a fix that seems to resolve the issue. I guess it only affects `main`.

**Review by orthagh — COMMENTED** [review]
A few remarks from a functional pov:


- First of all, real dashboard feels way better than the initial picture.

- widgets related to ticket status could be better distributed. if we present 12 months, recent should have opened statuses,  olders closed.
- we shouldn't present placeholder for mini-dashboard above tickets list. Firstly because the alert is not present, and also it induces an incoherence with the list below
- Thinking about the current way of having copies of existing dashboards: 
   They aims to present the feature and shouldn't have a life after that.
   And I'm not satisfied with the current process of moving to real data. A suggestion for a better flow (imo) : a single button on any example dashboard named "Disable demonstration" removing all fake dashboards, everywhere. The concept of defining the general/user dashboard could be too complex at the moment of the discovery of GLPI.

**Comment by orthagh** [comment]
Could you check this is working on firefox ? I have NS_BINDING_ABORTED messages, don't know why yet.
Also, the post url seems to target an url `dashboard.ajax.php` (the ajax. part doesn't exist)

**Comment by cconard96** [comment]
I don't see any issue unique to Firefox. I fixed the URL/button.

**Review by orthagh — CHANGES_REQUESTED** [review]
Firstly, functionally OK, this is what we wanted and it improves greatly the first experience with GLPI.

I know we changed a few times the direction of this recently but I'm not satisfied with the current storage currently. I mean technically.

Having a full copy in `glpi_dashboards_dashboards` (sufixed by `_placeholder`) and then on removing action, doing a 100lines method to sort out the database while we are in new installation (99% case), seems to be a complex implementation.
I don't see any scenarios where we may need placeholders for dashboard apart the new installation.

So, hopefully last iteration on this, and keeping the same exact experience (which is great, nothing more to say on this), could we move to a single global bool flag (maybe stored in glpi_configs) named, for example, is_demo_dashboards. Not storing anything more (no dashboard, no default dashboard for user or in general). This flag set to to true will show a dashboard stored in php classes directly. 

I can't pass too much time searching for a precise solution on this technical issue but we need a simple version of this pr. Don't hesitate to ping me on chat, and don't take me wrong, the current situation is more due to my indecision and changes of mind.

**Review by orthagh — APPROVED** [review]
Seems good on my side. Thank you for the switch on `glpi_configs`.
One last thing, there remains the pic previously used for demo. I think it's useless to keep it now.

**Review by cedric-anne — COMMENTED** [review]
Seems technically OK, but is it normal to be able to modify the "demo" dashboard ? The result is a mix between fake data and real data and it may be confusing.

**Comment by cconard96** [comment]
> Seems technically OK, but is it normal to be able to modify the "demo" dashboard ? The result is a mix between fake data and real data and it may be confusing.

As of the last few changes, there are no separate demo dashboards. There is a config option set during a fresh install that simply toggles the Provider used to supply data to the cards. When `is_demo_dashboards` is set to 1, the fake data provider is used for every card and that fake data provider should cover every case the regular provider does. So, being able to modify the default dashboards or add new ones seems OK to me at least. The alert stating they are viewing fake data will be shown on every dashboard until they disable the demo data.

**Review by cedric-anne — APPROVED** [review]
